### ISSUE TITLE
Java security manager and UDP SyslogAppender

### ISSUE THREAD

**Issue by vincentarnaud90** [elastic/elasticsearch#28996]
**Elasticsearch version** (`bin/elasticsearch --version`):
Version: 6.1.1, Build: bd92e7f/2017-12-17T20:23:25.338Z, JVM: 1.8.0_151

**Plugins installed**: []
None

**JVM version** (`java -version`):
openjdk version "1.8.0_151"
OpenJDK Runtime Environment (build 1.8.0_151-8u151-b12-1~deb9u1-b12)
OpenJDK 64-Bit Server VM (build 25.151-b12, mixed mode)

**OS version** (`uname -a` if on a Unix-like system):
Linux debian 9

**Description of the problem including expected versus actual behavior**:
I try to send elasticsearch logs into syslog using UDP. Nothing is sent because in org/elasticsearch/bootstrap/security.policy SocketPermission is missing "connect" (DatagramSocket.send needs it)
I tried to add "connect" in the policy file directly inside the jar and it seems to fix the issue.

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. Add an UDP SyslogAppender into log4j2.properties 
 2. Start elasticsearch with -Djava.security.debug=all to get relevant stacktrace

**Provide logs (if relevant)**:
```
access: access denied ("java.net.SocketPermission" "127.0.0.1:514" "connect,resolve")
java.lang.Exception: Stack trace
        at java.lang.Thread.dumpStack(Thread.java:1336)
        at java.security.AccessControlContext.checkPermission(AccessControlContext.java:462)
        at java.security.AccessController.checkPermission(AccessController.java:884)
        at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
        at java.lang.SecurityManager.checkConnect(SecurityManager.java:1051)
        at java.net.DatagramSocket.send(DatagramSocket.java:672)
        at org.apache.logging.log4j.core.net.DatagramOutputStream.flush(DatagramOutputStream.java:103)
        at org.apache.logging.log4j.core.appender.OutputStreamManager.flushDestination(OutputStreamManager.java:276)
        at org.apache.logging.log4j.core.appender.OutputStreamManager.write(OutputStreamManager.java:234)
        at org.apache.logging.log4j.core.appender.OutputStreamManager.write(OutputStreamManager.java:202)
        at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.writeByteArrayToManager(AbstractOutputStreamAppender.java:186)
        at org.apache.logging.log4j.core.appender.SocketAppender.directEncodeEvent(SocketAppender.java:451)
        at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.tryAppend(AbstractOutputStreamAppender.java:170)
        at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.append(AbstractOutputStreamAppender.java:161)
        at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
        at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
        at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
        at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
        at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:448)
        at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:433)
        at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:417)
        at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:403)
        at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:63)
        at org.apache.logging.log4j.core.Logger.logMessage(Logger.java:146)
        at org.apache.logging.log4j.spi.ExtendedLoggerWrapper.logMessage(ExtendedLoggerWrapper.java:217)
        at org.elasticsearch.common.logging.PrefixLogger.logMessage(PrefixLogger.java:102)
        at org.apache.logging.log4j.spi.AbstractLogger.tryLogMessage(AbstractLogger.java:2116)
        at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(AbstractLogger.java:2100)
        at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:1994)
        at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1966)
        at org.apache.logging.log4j.spi.AbstractLogger.info(AbstractLogger.java:1303)
        at org.elasticsearch.node.Node.<init>(Node.java:254)
        at org.elasticsearch.node.Node.<init>(Node.java:245)
        at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:212)
        at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:212)
        at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:322)
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:121)
        at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:112)
        at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86)
        at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
        at org.elasticsearch.cli.Command.main(Command.java:90)
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:92)
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:85)
```

**Comment by bleskes** [comment]
I don't think we want to relax the SM permissions. We should discuss and decide whether this is something we want to support (seems reasonable) and if so document how to do so.

**Comment by vincentarnaud90** [comment]
I made a mistake, it misses "connect" permission instead of "resolve". I'm correcting the initial post.

**Comment by rjernst** [comment]
We did a lot of work to remove connect permissions from core. I don't think we should add it back.

I think this goes back to us using log4j api directly. If it was wrapped, we could have logging implementations in modules, and relegate this permission only to that module.  But this (logging api issue) has been contentious in the past.

**Comment by vincentarnaud90** [comment]
Hi, I made some further testing. It seems to work fine using TCP instead of UDP, so we will see to switch protocol in syslog configuration.  
I also made a test inside a simple plugin (implementing an action filter which log something) to see if I can do something in there. Since it has it's own security policy, I though I might add the 'connect' permission. It didn't work as well (need to set permission inside elasticsearch jar) so I was wondering how "overriding" security policy works for plugins.

**Comment by rjernst** [comment]
Please see the [plugin author help docs](https://www.elastic.co/guide/en/elasticsearch/plugins/current/plugin-authors.html#plugin-authors-jsm) on security policies.

**Comment by vincentarnaud90** [comment]
Ok thanks, I should have read the doc, sorry.

**Comment by jasontedor** [comment]
> It seems to work fine using TCP instead of UDP, so we will see to switch protocol in syslog configuration.

If the TCP connection is terminated and Log4j needs to reconnect, this will not work because on the reconnect the security manager will be in place and reject the connection.

Additionally, I do not think you want to be synchronously writing to a network while logging. This can impact the entire system anywhere. We do not currently support the asynchronous appenders.

More broadly, I would say that we have other components in the stack that can forward logs to syslog. The advantage of these is that they are out of process from Elasticsearch, so if they fail or have issues they do not impact Elasticsearch itself.

In short, we do not support this, and it's possible that we will consider technical solutions that it make it clear that we do not support this.

**Comment by jasontedor** [comment]
Closing as won’t fix.

**Comment by vincentarnaud90** [comment]
Hi, thanks for the answer, we will see how we can do it outside elasticsearch then.

**Comment by jamshid** [comment]
Wow that's a big limitation -- elasticsearch does not support UDP or TCP syslog? 
Ideally it would be mentioned at https://www.elastic.co/guide/en/elasticsearch/reference/current/logging.html with pointers to alternative solutions.

I know this is not supported but if anyone else has to get an ancient Elasticsearch cluster with a working syslog configuration upgraded you can workaround this by adding a `grant` in a security policy that gets loaded like JAVA_HOME/lib/security/java.policy:
```
# Adds:
#    permission java.net.SocketPermission "mysyslog:514", "connect,resolve";
sed --in-place 's/permission java.net.SocketPermission "localhost:0", "listen";.*$/permission java.net.SocketPermission "localhost:0", "listen"; permission java.net.SocketPermission "'"${SYSLOG_HOST}:${SYSLOG_PORT}"'", "connect,resolve";/g' /etc/java/java-1.8.0-openjdk/java-1.8.0-openjdk-1.8.0.222.b10-0.el8_0.x86_64/lib/security/java.policy
```

**Comment by beorn-** [comment]
oh.
https://github.com/elastic/elasticsearch/issues/30849 explains it all !
